kong/db/dao/snis.lua

```diff
@@ -0,0 +1,158 @@
+      diff[len] = list1[i]
+    end
+  end
+
+  return diff
+end
+
+
+local _SNIs = {}
+
+
+-- Truthy if all the names on the list don't exist on the db
+function _SNIs:check_list_is_new(name_list)
+  -- when creating a new cert (no cert_id provided):
+  -- dont add the certificate or any names if we have a name conflict
+  -- its fairly inefficient that we have to loop twice over the datastore
+  -- but no support for OR queries means we gotsta!
+  for i=1, #name_list do
+    local name = name_list[i]
+    local row, err, err_t = singletons.db.snis:select_by_name(name)
+    if err then
+      return nil, err, err_t
+    end
+
+    if row then
+      -- Note: it could be that the name is not associated with any
+      -- certificate, but we don't handle this case. (for PostgreSQL
+      -- only, as C* requires a cert_id for its partition key).
+      local msg   = "SNI already exists: " .. name
+      local err_t = self.errors:conflicting_input(msg)
+      return nil, tostring(err_t), err_t
+    end
+  end
+
+  return 1
+end
+
+
+-- Truthy if all the names on the list don't exist on the db or exist but are
+-- associated to the given certificate
+function _SNIs:check_list_is_new_or_in_cert(cert_pk, name_list)
+  for i=1, #name_list do
+    local row, err, err_t = self:select_by_name(name_list[i])
+    if err then
+      return nil, err, err_t
+    end
+    if row and row.certificate.id ~= cert_pk.id then
+      local msg = "SNI '" .. row.name ..
+                  "' already associated with existing " ..
+                  "certificate (" .. row.certificate.id .. ")"
+      local err_t = self.errors:conflicting_input(msg)
+      return nil, tostring(err_t), err_t
+    end
+  end
+
+  return 1
+end
+
+
+-- Creates one instance of SNI for each name in name_list
+-- All created instances will be associated to the given certificate
+function _SNIs:insert_list(cert_pk, name_list)
+  for _, name in ipairs(name_list) do
+    local _, err, err_t = self:insert({
+      name         = name,
+      certificate  = cert_pk,
+    })
+    if err then
+      return nil, err, err_t
+    end
+  end
+
+  return 1
+end
+
+
+-- Deletes all SNIs on the given name list
+function _SNIs:delete_list(name_list)
+  local err_list = {}
+  local errors_len = 0
+  local first_err_t = nil
+  for i = 1, #name_list do
+    local ok, err, err_t = self:delete_by_name(name_list[i])
+    if not ok then
+      errors_len = errors_len + 1
+      err_list[errors_len] = err
+      first_err_t = first_err_t or err_t
+    end
+  end
+
+  if errors_len > 0 then
+    return nil, table.concat(err_list, ","), first_err_t
+  end
+
+  return 1
+end
+
+
+-- Returns the name list for a given certificate
+function _SNIs:list_for_certificate(cert_pk)
+  local name_list = {}
+  local rows, err, err_t = self:for_certificate(cert_pk)
+  if err then
+    return nil, err, err_t
+  end
+  for i = 1, #rows do
+    name_list[i] = rows[i].name
+  end
+
+  table.sort(name_list)
+  return name_list
```

Comment:
We should set the `cjson.empty_array_mt` metatable here, otherwise the Admin API will JSON-encode this as `{}` instead of `[]`.

Reply:
Done! (at the top of the function)